match/lib/match/storage/s3_storage.rb

```diff
@@ -0,0 +1,162 @@
+        # from all teams that use this bucket would be installed. This is not ideal, but
+        # unless the user provides a `team_id`, we can't know which one to use
+        # This only happens if `readonly` is activated, and no `team_id` was provided
+        @_folder_prefix ||= currently_used_team_id
+        if @_folder_prefix.nil?
+          # We use a `@_folder_prefix` variable, to keep state between multiple calls of this
+          # method, as the value won't change. This way the warning is only printed once
+          UI.important("Looks like you run `match` in `readonly` mode, and didn't provide a `team_id`. This will still work, however it is recommended to provide a `team_id` in your Appfile or Matchfile")
+          @_folder_prefix = "*"
+        end
+        return File.join(working_directory, @_folder_prefix)
+      end
+
+      # Call this method for the initial clone/download of the
+      # user's certificates & profiles
+      # As part of this method, the `self.working_directory` attribute
+      # will be set
+      def download
+        # Check if we already have a functional working_directory
+        return if @working_directory
+
+        # No existing working directory, creating a new one now
+        self.working_directory = Dir.mktmpdir
+
+        s3_client.find_bucket!(s3_bucket).objects.each do |object|
+          file_path = object.public_url.to_s.split("s3.amazonaws.com/").last
+          download_path = File.join(self.working_directory, file_path)
+
+          FileUtils.mkdir_p(File.expand_path("..", download_path))
+          UI.verbose("Downloading file from S3 '#{file_path}' on bucket #{self.s3_bucket}")
+
+          object.download_file(download_path)
+        end
+        UI.verbose("Successfully downloaded files from S3 to #{self.working_directory}")
+      end
+
+      # Returns a short string describing + identifing the current
+      # storage backend. This will be printed when nuking a storage
+      def human_readable_description
+        return "S3 Bucket [#{s3_bucket}] on region #{s3_region}"
+      end
+
+      def upload_files(files_to_upload: [], custom_message: nil)
+        # `files_to_upload` is an array of files that need to be uploaded to S3
+        # Those doesn't mean they're new, it might just be they're changed
+        # Either way, we'll upload them using the same technique
+
+        files_to_upload.each do |current_file|
+          # Go from
+          #   "/var/folders/px/bz2kts9n69g8crgv4jpjh6b40000gn/T/d20181026-96528-1av4gge/profiles/development/Development_me.mobileprovision"
+          # to
+          #   "profiles/development/Development_me.mobileprovision"
+          #
+
+          target_path = current_file.gsub(self.working_directory + "/", "")
+          UI.verbose("Uploading '#{target_path}' to S3 Storage...")
+
+          body = File.read(current_file)
+          acl = 'private'
+          s3_url = s3_client.upload_file(s3_bucket, target_path, body, acl)
```

Comment:
Does `upload_file` accept a path in addition to the contents? It would be nice not to have to read that if possible

Reply:
It currently accepts a body 😬